internal/targets/hash_bins.go

```diff
@@ -0,0 +1,119 @@
+package targets
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const MinDelegationHashPrefixBitLen = 1
+const MaxDelegationHashPrefixBitLen = 32
+
+// hexEncode formats x as a hex string. The hex string is left padded with
+// zeros to padWidth, if necessary.
+func hexEncode(x uint64, padWidth int) string {
+	// Benchmarked to be more than 10x faster than padding with Sprintf.
+	s := strconv.FormatUint(x, 16)
+	if len(s) >= padWidth {
+		return s
+	}
+	return strings.Repeat("0", padWidth-len(s)) + s
+}
+
+const bitsPerHexDigit = 4
+
+// numHexDigits returns is the number of hex digits required to encode
+// the given number of bits.
+func numHexDigits(numBits int) int {
+	// ceil(numBits / bitsPerHexDigit)
+	return ((numBits - 1) / bitsPerHexDigit) + 1
+}
+
+type HashBins struct {
+	rolePrefix string
+	bitLen     int
+
+	numBins           uint64
+	numPrefixesPerBin uint64
+
+	currIndex uint64
```

Comment:
Here is my suggestion:
instead of forming this as an iterator style, implement a getBin(index)/bin(index) that returns the bin struct with that index.

Reply:
Done. I was curious if this would have a big impact on memory allocs for a large number of hash bins, but a quick profiling shows that if you call `RoleName()` in the `HashBin` loop (as you almost always would), the `strconv.FormatUint` in `hexEncode` dominates allocs regardless of whether we use an iterator or index access. Tuning this is probably not so important right now (especially with TAP 15) -- was just curious.